### ISSUE TITLE
[question] backup rotate snap/backup snap rotate + rotate action

### ISSUE THREAD

**Issue by zilexa** [3coma3/btrfs-backup#17]
```
backup rotate snap - Will first apply the retention policy and then take a snapshot.

backup snap rotate - Will do the opposite, first take a snapshot and then check the snapshot repository against the policy.
```

"Take a snapshot" means copying files from source to a new btrfs snapshot on destination, right?
Isn't `rotate snap` the logical choice? In both cases (if snapshots have hit the rentention policy) at least one old snapshot will be deleted. Why have a `snap rotate` option, if a snapshot will be deleted anyway don't you want to create space on the drive first?

also.. I am no expert with rsync exclude options and I am trying to exclude some standard Linux/Windows/Mac folders and files, since I use my server to sync user content from several other devices. 
```
filters['/mnt/pool/Users/Local/Rudhra']="
- .filerun* # folders with ".filerun*" in the name should be excluded everywhere
- .Trash-* # folders with ".Trash-*" in the name should be excluded everywhere
- Thumbs.db # files with this name should be excluded everywhere
- RECYCLE.BIN # folders with this name should be excluded everywhere
- .Spotlight-V100 # folders with this name should be excluded everywhere
- .TemporaryItems # folders with this name should be excluded everywhere
- .Trashes # folders with this name should be excluded everywhere
- .fseventsd # folders with this name should be excluded everywhere
- .DS_Store # folders with this name should be excluded everywhere
```
Just wondering if there is special syntax needed, I think it is OK like this. 

Perhaps also a suggestion if you really have spare time: allow to have 1 config file, with 1 destination and and one source folder, letting the script create subfolders (that will each contain the snapshots folder) in the destination for each folder it detects (1 level) within the source folder.
This way it is even easier to backup content on a NAS used by multiple users. When a user is added or removed, no need to add/remove .conf files.

**Comment by 3coma3** [comment]
Hi,

> "Take a snapshot" means copying files from source to a new btrfs snapshot on destination, right?

That's correct 

> Isn't `rotate snap` the logical choice? In both cases (if snapshots have hit the rentention policy) at least one old snapshot will be deleted. Why have a `snap rotate` option, if a snapshot will be deleted anyway don't you want to create space on the drive first?

I actually use snap rotate, because I prefer to include the last snapshot while computing rotation. Note that ```rotate``` will not always _actually_ rotate, that depends on the policy and amount of snapshots at each level in that particular application.

Moreover, even if you are sure that one ```rotate``` application will actually mark snapshots for rotation... _in general_ this doesn't mean that you will end up freeing a lot of space, this actually depends on the actual size and free space on the destination volume, and whether there are big enough differences in the snapshots that would be rotated to be something meaningful or not.

Precisely for these ambiguities is that I think that there isn't one sequence that is right for every case. Besides, if more actions are added in the future that can benefit from the free ordering and sequencing (I call this "command queueing" in the context of this script), then the feature is already here.  

> also.. I am no expert with rsync exclude options and I am trying to exclude some standard Linux/Windows/Mac folders and files, since I use my server to sync user content from several other devices.
> 
> ```
> filters['/mnt/pool/Users/Local/Rudhra']="
> - .filerun* # folders with ".filerun*" in the name should be excluded everywhere
> - .Trash-* # folders with ".Trash-*" in the name should be excluded everywhere
> - Thumbs.db # files with this name should be excluded everywhere
> - RECYCLE.BIN # folders with this name should be excluded everywhere
> - .Spotlight-V100 # folders with this name should be excluded everywhere
> - .TemporaryItems # folders with this name should be excluded everywhere
> - .Trashes # folders with this name should be excluded everywhere
> - .fseventsd # folders with this name should be excluded everywhere
> - .DS_Store # folders with this name should be excluded everywhere
> ```
> 
> Just wondering if there is special syntax needed, I think it is OK like this.

Yes, I also think it's ok. In case you want a folder to appear in the snapshots (but not its contents), the syntax is: ```- folder/*```

> Perhaps also a suggestion if you really have spare time: allow to have 1 config file, with 1 destination and and one source folder, letting the script create subfolders (that will each contain the snapshots folder) in the destination for each folder it detects (1 level) within the source folder.
> This way it is even easier to backup content on a NAS used by multiple users. When a user is added or removed, no need to add/remove .conf files.

I don't think it would be a good solution to support such a specific case. This is the type of feature that seems innocent enough but when you add another 10 specific scenarios, the code is already unmaintainable. A different kind of situation would be with things that are very general and would have a bigger impact, like for example:

- BTRFS send/receive (like we talked about before)
- ZFS support
- new rotation handlers
- improved output
- a better plugin oriented architecture, even

What I would do here is to add a feature that is general/logical/commonplace enough as to merit the addition because it would help not only in your specific case, but also in others that we cannot foresee. Concretely, the script could accept the source and destination parameters, either (or both) from command line and/or environment variables.

This way you could do something like this for your problem:

1. Have a configuration that doesn't specify source and destination (or that specifies them, but a theoretical "override from command line/env" is supported so it doesn't matter).
2. Instead of having something like this in cron:

```/usr/local/backup -c myconfig.conf snap rotate```

Have something like:

```
for dir in "$source/*"; do
  [[ -d "$dir" ]] &&
  /usr/local/backup -c myconfig.conf --source "$dir" --destination "/media/me/snapsroot/$(basename "$dir")" snap rotate
done
```

(assuming that the root of all the snapshots is located at /media/me/snapsroot)

So basically it would be possible to add a little shell spice and have the script to behave when passing the source and destination in this way. It's likely that when using a command line switch it would be possible to pass only a single source directory, but when using environment variables you could still use an array, like currently in the script.

What do you think?

**Comment by 3coma3** [comment]
Closing for now, feel free to reopen
